src/lando/pulse/pulse.py

```diff
@@ -0,0 +1,112 @@
+                + "This should not be the case in non-local deployments. "
+                + str(settings.PULSE_HOST)
+            )
+            if settings.ENVIRONMENT not in ["test", "local", "suite"]:
+                logger.warning(message, extra={"PULSE_HOST": settings.PULSE_HOST})
+            if settings.ENVIRONMENT.startswith("prod"):
+                # XXX: we should verify this much earlier on
+                raise RuntimeError(message)
+
+        connection = kombu.Connection(
+            hostname=settings.PULSE_HOST,
+            port=settings.PULSE_PORT,
+            userid=settings.PULSE_USER,
+            password=settings.PULSE_PASSWORD,
+            connect_timeout=100,
+            ssl=settings.PULSE_SSL,
+        )
+        connection.connect()
+
+        ex = kombu.Exchange(settings.PULSE_EXCHANGE, type="direct")
+        queue = kombu.Queue(
+            name=settings.PULSE_QUEUE,
+            exchange=settings.PULSE_EXCHANGE,
+            routing_key=settings.PULSE_ROUTING_KEY,
+            durable=True,
+            exclusive=False,
+            auto_delete=False,
+        )
+        queue(connection).declare()
+
+        producer = connection.Producer(
+            exchange=ex, routing_key=settings.PULSE_ROUTING_KEY, serializer="json"
+        )
+        return producer
+
+    def notify_push(self, push: Push):
+        message = self.pulse_message_for_push(push)
+
+        # XXX: make a separate notification worker that loops around un-notified
+        # Push entries.
+        logger.info(f"Sending {message} ...")
+        self.producer.publish(
+            message,
+            retry=True,
+            retry_policy={
+                "interval_start": 0,  # First retry immediately,
+                "interval_step": 2,  # then increase by 2s for every retry.
+                "interval_max": 30,  # but don't exceed 30s between retries.
+                "max_retries": 30,  # give up after 30 tries.
+            },
+        )
+
+        push.notified = True
+        push.save()
+
+    @classmethod
+    def pulse_message_for_push(cls, push: Push):
+        push_data = model_to_dict(push)
+
+        branches = {}
+        if push.commits.count():
+            commit = push.commits.latest()
+            branches = {push_data["branch"]: commit.hash}
+
+        # XXX: to be implemented
+        tags = {}
+
+        if not branches and not tags:
+            logger.error(
+                f"Push {push.push_id} does not contain either branches or tags"
+            )
+
+        message = {
+            "payload": {
+                "type": "push",
+                "repo_url": push_data["repo_url"],
+                "branches": branches,
+                "tags": {},
+                "time": push.datetime.strftime("%s"),
+                "push_id": push_data["push_id"],
+                "user": push_data["user"],
+                # XXX
+                "push_json_url": "FIXME",
+                "push_full_json_url": "FIXME",
```

Comment:
I guess we don't have a URL since we don't have the web view yet. :sweat_smile:  Can we add a bug number?

Reply:
Added (: